jpx/src/main/java/io/jenetics/jpx/GPX.java

```diff
@@ -1137,10 +1138,10 @@ public GPX read(final InputStream input)
 					reader.next();
 					return _reader.read(reader, _mode == Mode.LENIENT);
 				} else {
-					throw new IOException("No 'gpx' element found.");
+					throw new InvalidObjectException("No 'gpx' element found.");
 				}
 			} catch (XMLStreamException e) {
-				throw new IOException(e);
+				throw new InvalidObjectException("Invalid 'gpx' input.");
```

Comment:
We have to catch also an `IllegalArgumentException` here:
```java
catch (IllegalArgumentException e) {
    throw (InvalidObjectException)new InvalidObjectException(e.getMessage())
        .initCause(e);
}
```

The `fromString` method could also be optimized.
```java
public GPX fromString(final String xml) {
	final byte[] bytes = xml.getBytes();
	final ByteArrayInputStream in = new ByteArrayInputStream(bytes);
	try {
		return read(in);
	} catch (InvalidObjectException e) {
		if (e.getCause() instanceof IllegalArgumentException) {
			throw (IllegalArgumentException)e.getCause();
		}
		throw new IllegalArgumentException(e);
	} catch (IOException e) {
		throw new IllegalArgumentException(e);
	}
}
```

Reply:
Yes, having unit tests would be a great idea. :+1: